src/rule_definition_tools.jl

```diff
@@ -392,7 +396,70 @@ function _nondiff_rrule_expr(__source__, primal_sig_parts, primal_invoke)
 end
 
 
-###########
+############################################################################################
+# @opt_out
+
+"""
+    @opt_out frule([config], _, f, args...)
+    @opt_out rrule([config], f, args...)
+
+This allows you to opt-out of a `frule` or `rrule` by providing a more specific method,
+that says to use the AD system, to solver it.
+
+For example, consider some function `foo(x::AbtractArray)`.
+In general, you know a efficicent and generic way to implement it's `rrule`.
+You do so, (likely making use of [`ProjectTo`](@ref)).
+But it actually turns out that for some `FancyArray` type it is better to let the AD do it's
+thing.
+
+Then you would write something like:
+```julia
+function rrule(::typeof(foo), x::AbstractArray)
+    foo_pullback(ȳ) = ...
+    return foo(x), foo_pullback
+end
+
+@opt_out rrule(::typeof(foo), ::FancyArray)
+```
+
+This will generate a [`rrule`](@ref) that returns `nothing`,
+and will also add a similar entry to [`ChainRulesCore.no_rrule`](@ref).
+
+Similar applies for [`frule`](@ref) and [`ChainRulesCore.no_frule`](@ref)
+"""
+macro opt_out(expr)
+    no_rule_target = _no_rule_target_rewrite!(deepcopy(expr))
+
+    return @strip_linenos quote
+        $(esc(no_rule_target)) = nothing
+        $(esc(expr)) = nothing
+    end
+end
+
+function _no_rule_target_rewrite!(call_target::Symbol)
+    return if call_target == :rrule
+        :(ChainRulesCore.no_rrule)
+    elseif call_target == :frule
+        :(ChainRulesCore.no_frule)
+    else
+        error("Unexpected opt-out target. Exprected frule or rrule, got: $call_target")
+    end
+end
+_no_rule_target_rewrite!(qt::QuoteNode) = _no_rule_target_rewrite!(qt.value)
+function _no_rule_target_rewrite!(expr::Expr)
+    length(expr.args)===0 && error("Malformed method expression. $expr")
+    if expr.head === :call || expr.head === :where
```

Comment:
I can't figure out what the initial expression would be that hits the `:where` branch someplace along the execution?

Reply:
`@opt_out rrule(typeof(foo), ::T, ::T) where T`

```julia
julia> dump(:(rrule(typeof(foo), ::T, ::T) where T))
Expr
  head: Symbol where
  args: Array{Any}((2,))
    1: Expr
      head: Symbol call
      args: Array{Any}((4,))
        1: Symbol rrule
        2: Expr
          head: Symbol call
          args: Array{Any}((2,))
            1: Symbol typeof
            2: Symbol foo
        3: Expr
          head: Symbol ::
          args: Array{Any}((1,))
            1: Symbol T
        4: Expr
          head: Symbol ::
          args: Array{Any}((1,))
            1: Symbol T
    2: Symbol T
```